File: Algorithm/BOJ/Java/Dynamic Programming/11052.java
package boj;

import java.util.Scanner;

public class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		
		int N = sc.nextInt();
		
		int[] cards = new int[N + 1];
		cards[0] = 0;
		for(int i = 1; i <= N; i++) {			
			cards[i] = sc.nextInt(); 
		}
		
		System.out.println(getMaxPrice(N, cards));
	}
	
	public static int getMaxPrice(int N, int cards[]) {
		int[] answer = new int[N + 1];
		
		for(int i = 1; i < N + 1; i++) {
			for(int j = 1; j < i + 1; j++) {
				answer[i] = getMax(answer[i], answer[i - j] + cards[j]);						
			}
		}
		
		return answer[N];
	}
	
	public static int getMax(int a, int b) {
		if (a >= b) {
			return a;
		}
		else {
			return b;				
		}		
	}
}

File: Algorithm/BOJ/Go/Dynamic Programming/11052.go
// 11052
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)

	var N int
	fmt.Fscanln(reader, &N)

	cards := make([]int, N+1)
	for i := 1; i <= N; i++ {
		fmt.Fscanf(reader, "%d ", &cards[i])
	}

	fmt.Println(getMaxPrice(N, &cards))
}

func getMaxPrice(N int, cards *[]int) int {
	answer := make([]int, N+1)

	for i := 1; i < N+1; i++ {
		for j := 1; j < i+1; j++ {
			answer[i] = getMax(answer[i], answer[i-j]+(*cards)[j])
		}
	}

	return answer[N]
}

func getMax(a int, b int) int {
	if a >= b {
		return a
	} else {
		return b
	}
}
